src/main/java/com/comphenix/protocol/wrappers/WrappedParticle.java

```diff
@@ -1,169 +1,184 @@
+				.newBuilder()
+				.requireModifier(Modifier.STATIC)
+				.returnTypeExact(Particle.class)
+				.parameterExactType(MinecraftReflection.getParticleParam())
+				.build();
+		toBukkit = Accessors.getMethodAccessor(fuzzy.getMethod(contract));
+
+		contract = FuzzyMethodContract
+				.newBuilder()
+				.requireModifier(Modifier.STATIC)
+				.returnTypeExact(MinecraftReflection.getParticleParam())
+				.parameterCount(2)
+				.build();
+		toNMS = Accessors.getMethodAccessor(fuzzy.getMethod(contract));
+
+		Class<?> cbData = MinecraftReflection.getCraftBukkitClass("block.data.CraftBlockData");
+		fuzzy = FuzzyReflection.fromClass(cbData);
+		contract = FuzzyMethodContract
+				.newBuilder()
+				.requireModifier(Modifier.STATIC)
+				.returnTypeExact(cbData)
+				.parameterExactArray(MinecraftReflection.getIBlockDataClass())
+				.build();
+		toCraftData = Accessors.getMethodAccessor(fuzzy.getMethod(contract));
+	}
+
+	private final Particle particle;
+	private final T data;
+	private final Object handle;
+
+	private WrappedParticle(Object handle, Particle particle, T data) {
+		this.handle = handle;
+		this.particle = particle;
+		this.data = data;
+	}
+
+	/**
+	 * @return This particle's Bukkit type
+	 */
+	public Particle getParticle() {
+		return particle;
+	}
+
+	/**
+	 * Gets this Particle's Bukkit/ProtocolLib data. The type of this data depends on the
+	 * {@link #getParticle() Particle type}. For Block particles it will be {@link WrappedBlockData},
+	 * for Item crack particles, it will be an {@link ItemStack}, and for redstone particles it will
+	 * be {@link Particle.DustOptions}
+	 *
+	 * @return The particle data
+	 */
+	public T getData() {
+		return data;
+	}
+
+	/**
+	 * @return NMS handle
+	 */
+	public Object getHandle() {
+		return handle;
+	}
+
+	public static WrappedParticle fromHandle(Object handle) {
+		ensureMethods();
+
+		Particle bukkit = (Particle) toBukkit.invoke(null, handle);
+		Object data = null;
+
+		switch (bukkit) {
+			case BLOCK_CRACK:
+			case BLOCK_DUST:
+			case FALLING_DUST:
+				data = getBlockData(handle);
+				break;
+			case ITEM_CRACK:
+				data = getItem(handle);
+				break;
+			case REDSTONE:
+				data = getRedstone(handle);
+				break;
+			default:
+				break;
+		}
+
+		return new WrappedParticle<>(handle, bukkit, data);
+	}
+
+	private static WrappedBlockData getBlockData(Object handle) {
+		return new StructureModifier<>(handle.getClass())
+				.withTarget(handle)
+				.withType(MinecraftReflection.getIBlockDataClass(), BukkitConverters.getWrappedBlockDataConverter())
+				.read(0);
+	}
+
+	private static Object getItem(Object handle) {
+		return new StructureModifier<>(handle.getClass())
+				.withTarget(handle)
+				.withType(MinecraftReflection.getItemStackClass(), BukkitConverters.getItemStackConverter())
+				.read(0);
+	}
+
+	private static Object getRedstone(Object handle) {
+		int r, g, b;
+		float alpha;
+
+		if (MinecraftVersion.FEATURE_PREVIEW_UPDATE.atOrAbove()) {

```

Comment:
there's probably a fuzzy class contract that could cover these 2 implementations of vector3f